src/YDialogSpy.cc

```diff
@@ -278,70 +474,427 @@ void fillTree( YWidgetTreeItem * 		parent,
+        else if (type == "ReplacePoint")
+            editWidget(f->createReplacePoint(widget));
+        else if (type == "Right")
+            editWidget(f->createRight(widget));
+        else if (type == "RichText")
+            editWidget(f->createRichText(widget, "This is a <b>RichText</b>."));
+        else if (type == "SelectionBox")
+            editWidget(f->createSelectionBox(widget, "Selection Box"));
+        else if (type == "Table")
+        {
+            YPopupInternal::StringArray arr { "", "", "" };
+
+            // abort adding if Cancel has been pressed
+            if (YPopupInternal::editStringArray(arr, "Table Columns"))
+            {
+                auto header = new YTableHeader();
+
+                // access by reference
+                for(auto&& str: arr) header->addColumn(str);
+
+                editWidget(f->createTable(widget, header));
+            }
+        }
+        else if (type == "Top")
+            editWidget(f->createTop(widget));
+        else if (type == "Tree")
+            editWidget(f->createTree(widget, "Tree"));
+        else if (type == "VBox")
+            editWidget(f->createVBox(widget));
+        else if (type == "VSpacing")
+            editWidget(f->createVSpacing(widget));
+        else if (type == "VStretch")
+            editWidget(f->createVStretch(widget));
+        else
+        {
+            YPopupInternal::message(
+                "Adding \"" + type + "\" widget type is not supported.");
+            return;
+        }
+
+        targetDialogUpdated();
     }
-    catch ( YUIException & exception )
+    catch( const YUIException & exception )
+    {
+        YPopupInternal::message("Could not add a new widget:\n"
+            + exception.msg());
+    }
+}
+
+/**
+ * Refresh the target dialog after modifying it.
+ */
+void YDialogSpyPrivate::targetDialogUpdated()
+{
+    // redraw the target dialog
+    targetDialog->recalcLayout();
+
+    // refresh the spy dialog
+    widgetTree->deleteAllItems();
+    fillWidgetTree(targetDialog, widgetTree);
+}
+
+/**
+ * Refresh button states in the main spy dialog
+ */
+void YDialogSpyPrivate::refreshButtonStates()
+{
+    auto widget = selectedWidget();
+    auto parent = widget ? widget->parent() : nullptr;
+
+    // Enable the moving buttons ony when the selected widget is inside
+    // a VBox/HBox container, set the labels according to stacking direction.
+    if (widget && parent && isBox(parent))
+    {
+        upButton->setEnabled(widget != parent->firstChild());
+        upButton->setLabel(isVBox(parent) ? "⬆ Up" : "⬅ Left");
+        downButton->setEnabled(widget != parent->lastChild());
+        downButton->setLabel(isVBox(parent) ? "⬇ Down" : "➡ Right");
+    }
+    else
     {
-	YUI_CAUGHT( exception );
+        upButton->setEnabled(false);
+        downButton->setEnabled(false);
     }
+
+    // TODO: Enable the [Add] menu button only when a widget can be added
+    // inside the current widget (i.e. it is a container).
```

Comment:
IIUC this can be deduced if we ask for the widget's children manager and check if it is not a YSingleWidgetChildManager or a YWidgetChildrenRejector : https://doc.opensuse.org/projects/libyui/HEAD/classYChildrenManager.html